File: moq-web/src/internal/message/announce_interest.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface AnnounceInterestMessageInit {
	prefix?: string;
	excludeHop?: number;
}

export class AnnounceInterestMessage {
	prefix: string;
	excludeHop: number;

	constructor(init: AnnounceInterestMessageInit = {}) {
		this.prefix = init.prefix ?? "";
		this.excludeHop = init.excludeHop ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.string(this.prefix);
		e.varint(this.excludeHop);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);

		this.prefix = d.string();
		this.excludeHop = d.varint();

		return undefined;
	}
}

File: moqt/internal/message/announce_interest.go
package message

import (
	"io"
)

/*
 *	ANNOUNCE_INTEREST Message {
 *	  Message Length (i)
 *	  Broadcast Path Prefix (s),
 *	  Exclude Hop (i),
 *	}
 */
type AnnounceInterestMessage struct {
	BroadcastPathPrefix string
	ExcludeHop          uint64
}

func (aim AnnounceInterestMessage) Len() int {
	return StringLen(aim.BroadcastPathPrefix) + VarintLen(aim.ExcludeHop)
}

func (aim AnnounceInterestMessage) Encode(w io.Writer) error {
	msgLen := aim.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteString(b, aim.BroadcastPathPrefix)
	b, _ = WriteVarint(b, aim.ExcludeHop)

	_, err := w.Write(b)

	return err
}

func (aim *AnnounceInterestMessage) Decode(src io.Reader) error {
	num, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, num)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	aim.BroadcastPathPrefix = str
	b = b[n:]

	excludeHop, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	aim.ExcludeHop = excludeHop
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}
